core/src/main/scala/kafka/server/metadata/PartitionMetadataProcessor.scala

```diff
@@ -97,20 +98,29 @@ import scala.jdk.CollectionConverters._
  *                      when partition counts change
  * @param replicaManager the replica manager
  * @param txnCoordinator the transaction coordinator
+ * @param configHandlers the config handlers
  */
 class PartitionMetadataProcessor(kafkaConfig: KafkaConfig,
                                  clusterId: String,
                                  metadataCache: MetadataCache,
                                  groupCoordinator: GroupCoordinator,
                                  quotaManagers: QuotaFactory.QuotaManagers,
                                  replicaManager: ReplicaManager,
-                                 txnCoordinator: TransactionCoordinator) extends BrokerMetadataProcessor with Logging {
+                                 txnCoordinator: TransactionCoordinator,
+                                 configHandlers: Map[ConfigResource.Type, ConfigHandler]) extends BrokerMetadataProcessor with Logging {
   // used only for onLeadershipChange()
   private val apisUtils = new ApisUtils(null, None, null, null, Some(groupCoordinator), Some(txnCoordinator))
 
   // visible for testing
   private[metadata] var brokerEpoch: Long = -1
 
+  // Define an immutable map for configs.  Every config change effectively results in this var being set to a new
+  // map.  Every instance is immutable, and updates (always performed by the same thread) replace the var with a
+  // completely new value. This means reads (which can occur from any thread) need to grab the value of this var
+  // (into a val) ONCE and retain that read copy for the duration of their operation. Multiple reads of the var
+  // risk getting different values.
+  @volatile private var configMap: Map[ConfigResource, Map[String, String]] = Map.empty
```

Comment:
I think this is too inefficient.  Imagine if we have a million partitions, all with configs.  We're going to create a new million-entry map each time the configuration changes?  Mutexes and a mutable map are more sensible here.

Reply:
ConfigResource is only at the topic or broker level -- not at the partition level.  Or am I missing something?

Reply:
ok, fair, but you still need to fix this since we can have a lot of topics